### ISSUE TITLE
[HOLD for payment 2023-10-02] [HOLD for payment 2023-07-24] [$1000] Failed Room Termination: User Continues Chatting in Android App After Leaving Room on Web

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#21518]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Login with User A ( mobile number) > create a workspace with user B and C
2. Create a room in public mode
3. Send the public room link or QR code to User B
4. Open User B in android as well as Desktop chrome and join the room across web and app using link & QR code
5. From User B, send messages to the room
6. From User B, Leave the room from web
7. Verify if User B can still send messages from android app
## Expected Result:
When a user leaves a room on the web version, their participation in the room should be terminated across all platforms, including the Android app. They should not be able to send messages or interact within the room.
## Actual Result:
After a user leaves a room on the web version, they can still send messages and participate in the chat room using the Android app. This behaviour contradicts the expected result and allows the user to continue engaging with the room despite their departure on the web.
## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.32-5
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/aa704c3f-7c19-4450-a800-ac27ce1d5ae9

https://github.com/Expensify/App/assets/43996225/330ae59c-16e6-4be6-8164-dfbacfbd1ec6

https://github.com/Expensify/App/assets/43996225/b6d6ed30-546c-40ec-aa9b-3dfd284d0944

**Expensify/Expensify Issue URL:**
**Issue reported by:** @avi-shek-jha
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1687048129225459

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~012dc182ad93ca216b</li>
        <li>Upwork Job ID: 1674172420971741184</li>
        <li>Last Price Increase: 2023-07-05</li>
<li>Automatic offers: </li>
<ul>
<li>wildan-m | Contributor | 27225915</li>
<li>avi-shek-jha | Reporter | 27225918</li>
</ul></li>
    </ul>
</details>

**Comment by mkarkachov** [comment]
screencast from [bug report](https://expensify.slack.com/archives/C049HHMV9SM/p1686850018180709)
https://github.com/Expensify/App/assets/1360242/9f58a028-f357-4929-9f04-461006082b16

**Comment by mkarkachov** [comment]
@kavimuru I posted converted video (changed extension) as you [requested](https://expensify.slack.com/archives/C049HHMV9SM/p1687707397155209?thread_ts=1686850018.180709&cid=C049HHMV9SM)

**Comment by johncschuster** [comment]
Working on reproducing this behavior this afternoon.

**Comment by jeet-dhandha** [comment]
@allroundexperts why am i mentioned here 🤔 ?

**Comment by nishancx** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Failed Room Termination: User Continues Chatting in Android App After Leaving Room on Web

### What is the root cause of that problem?
The nature of public rooms is that even after leaving a room, when user visits the room with the room link, the user joins the room. So, we need make sure that user leaving a room from device A doesn't make device B visit that room again. When user leaves public room form device A, onyx data for that report is replaced through `pusher` with `null` value, this causes the user B to re-fetch room data which in turn causes user to rejoin the room.

### What changes do you think we should make in order to solve the problem?
In `ReportScreen` -> `componentDidUpdate`, `fetchReportIfNeeded()` function is called which re-fetches the room and makes user rejoin the room. An `if` condition should be added so that `fetchReportIfNeeded()` isn't called if `this.props.report.reportId` doens't exist ie. the report's value is `null`.

**Comment by wildan-m** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Failed Room Termination: User Continues Chatting in Android App After Leaving Room on Web

### What is the root cause of that problem?

The public room is re-opened on the second device (android) so the user will re-joined

### What changes do you think we should make in order to solve the problem?

IMO, adding conditional logic to skip `fetchReportIfNeeded()` is not enough, since the report will still persist and the report can be reopened by another trigger e.g. `onVisibilityChange`

I'd propose adding an early return if the report is deleted.

We can change this code:

https://github.com/Expensify/App/blob/7e908b17565e7bf659309f4b4d95141e70094ab9/src/pages/home/ReportScreen.js#L162-L170

To:

```
        const onyxReportID = this.props.report.reportID;
        const routeReportID = getReportID(this.props.route);

        if(prevProps.report.reportID && !onyxReportID){
             return;
        }

        // If you already have a report open and are deeplinking to a new report on native,
        // the ReportScreen never actually unmounts and the reportID in the route also doesn't change.
        // Therefore, we need to compare if the existing reportID is the same as the one in the route
        // before deciding that we shouldn't call OpenReport.
        if (onyxReportID === prevProps.report.reportID && (!onyxReportID || onyxReportID === routeReportID)) {
            return;
        }
```

This will make the user doesn't have access to the page immediately.

Or if we want to follow [the behavior in `leaveRoom` ](https://github.com/Expensify/App/blob/7e908b17565e7bf659309f4b4d95141e70094ab9/src/libs/actions/Report.js#L1729)function, we can change the `return;` with `Report.navigateToConciergeChat();`.

         

### What alternative solutions did you explore? (Optional)
N/A

**Comment by johncschuster** [comment]
@allroundexperts can you take a look at the above proposals when you get a moment? Do we need to keep looking?

**Comment by wildan-m** [comment]
@nishancx, how can we refresh the page in the native app?

**Comment by wildan-m** [comment]
@allroundexperts sure! [Proposal Updated](https://github.com/Expensify/App/issues/21518#issuecomment-1621007692).

**Comment by wildan-m** [comment]
@allroundexperts we can call goback before navigate to concierge, like we did in conciergepage

**Comment by johncschuster** [comment]
> @johncschuster Should we redirect the user to concierge chat if he's removed from the room? I think it makes sense given we are doing the same in `leaveRoom`.

@allroundexperts that makes sense to me!

**Comment by nishancx** [comment]
> @nishancx, how can we refresh the page in the native app?

Sorry for the confusion, I was talking about two concurrent sign-ins in web, as this issue can be replicated in not only web-android but also web-web

**Comment by johncschuster** [comment]
@allroundexperts sorry for the additional bump. Do you feel we have what we need? If not, let's take this to #expensify-open-source.

**Comment by wildan-m** [comment]
@johncschuster @allroundexperts Proposal [Updated](https://github.com/Expensify/App/issues/21518#issuecomment-1621007692).

**Comment by wildan-m** [comment]
@allroundexperts

Proposal [Updated](https://github.com/Expensify/App/issues/21518#issuecomment-1621007692).
I've found a possible regression when the user re-joined the public room right after leaving the room.

**Comment by johncschuster** [comment]
@allroundexperts, looks like there's been one more update to @wildan-m's proposal. Can you give it another pass through? If you're happy with it, I'll get the assignments going.

**Comment by johncschuster** [comment]
@allroundexperts can you complete the BZ checklist above when you get a moment?

**Comment by joelbettner** [comment]
Any more word on this?

**Comment by johncschuster** [comment]
Sure! I'll assign someone internal to this. Thanks, @allroundexperts!

**Comment by johncschuster** [comment]
@luacmartins heads up! @allroundexperts suggested in [this comment](https://github.com/Expensify/App/issues/21518#issuecomment-1659525004) that we need to take this internal. He also provided a Slack link for context on the discussion.

**Comment by luacmartins** [comment]
thanks for the ping @johncschuster! I left a comment in [this thread](https://github.com/Expensify/App/pull/23051#discussion_r1282253827)

**Comment by luacmartins** [comment]
Discussing in this thread - https://expensify.slack.com/archives/C01GTK53T8Q/p1691111512068339

**Comment by luacmartins** [comment]
Still discussing in thread. Came up with a new approach today!

**Comment by wildan-m** [comment]
Just resolved another conflict. @johncschuster seems @allroundexperts is busy at the moment, can someone help review [this follow-up PR](https://github.com/Expensify/App/pull/24407)?

Context about changes: https://expensify.slack.com/archives/C01GTK53T8Q/p1691111512068339

**Comment by wildan-m** [comment]
@allroundexperts , sure, thanks for your feedback

**Comment by wildan-m** [comment]
@allroundexperts latest main has been merged

**Comment by wildan-m** [comment]
@allroundexperts latest main has been merged

**Comment by wildan-m** [comment]
@allroundexperts latest main merged and the new conflict has been resolved. any updates?

**Comment by wildan-m** [comment]
seems @MelvinBot tagged the wrong person

**Comment by wildan-m** [comment]
@johncschuster, @luacmartins,

Have you been following the discussion in [this post](https://github.com/Expensify/App/issues/28522)?

In my opinion, the issue mentioned in that post is not directly related to [my PR](https://github.com/Expensify/App/pull/24407). This has been confirmed both [here](https://github.com/Expensify/App/issues/28522#issuecomment-1742044696) and [here](https://github.com/Expensify/App/issues/28522#issuecomment-1742921730).

There's another issue stemming from the implications of [this PR](https://github.com/Expensify/App/pull/27127).

This issue alters the expected behavior of the original problem:

Before: Leaving all types of rooms, including threads, would navigate to the concierge.
After: Leaving the thread room will navigate to the parent thread, while other room types will navigate to the concierge.

Since there's been a change in the requirements, I, along with @allroundexperts, agree that it's not a regression.

The solution to this change might be straightforward. We can add the following code snippet to `ReportScreen` as suggested [here](https://github.com/Expensify/App/issues/28522#issuecomment-1742506182):

```
if (report.parentReportID) {
    Navigation.navigate(ROUTES.REPORT_WITH_ID.getRoute(report.parentReportID), CONST.NAVIGATION.TYPE.FORCED_UP);
    return;
}
```

However, implementing this change will necessitate more extensive testing since `ReportScreen` has a significant impact on various features, and I believe it may require additional adjustments.

In my opinion, it's better to finalize this issue and create a separate one to address the requirement changes.

**Comment by luacmartins** [comment]
I wasn't aware of that discussion, but I just caught up, thanks for the link. I agree that it's not a regression and that we should close out this issue once payment is issued. I also agree that we should create a new issue for the navigation requirement change. Wanna report the issue and I can create it?

**Comment by wildan-m** [comment]
@luacmartins sure, I'll post it to expensify-bugs, please wait

**Comment by wildan-m** [comment]
@luacmartins posted [here](https://expensify.slack.com/archives/C049HHMV9SM/p1696609581710049).

**Comment by wildan-m** [comment]
@luacmartins good news! the requirement changes have been resolved by [this PR](https://github.com/Expensify/App/pull/28702).

**Comment by luacmartins** [comment]
Cool, seems like we're all set then and just need to process payment for this issue

**Comment by wildan-m** [comment]
👍

**Comment by wildan-m** [comment]
Bump @luacmartins, @johncschuster

**Comment by wildan-m** [comment]
Bump @johncschuster, @luacmartins

**Comment by johncschuster** [comment]
Ok, the job seems to not want to open back up. I'm going to create an issue just for payment for this. Standby.

**Comment by johncschuster** [comment]
@wildan-m can you please go to the linked issue above and comment on it so I can assign it to you? That should get the job offer sent to you correctly.

**Comment by avi-shek-jha** [comment]
Hello, as a reporter, I only received an offer for 50 dollars. I am aware of the new policies. But, I reported this on June 17 and the issue was created on June 25 almost 2 months before the new policies took place. Can anyone please send me the offer link with corrected amount of 250 USD?

**Comment by johncschuster** [comment]
Hi @avi-shek-jha, thanks for calling that out. I will correct that in Upwork.

**Comment by johncschuster** [comment]
@avi-shek-jha, I've corrected that in Upwork. Can you please accept the offer? I'll take care of payment once you've done that.

**Comment by avi-shek-jha** [comment]
Thank you John. I appreciate it. I have accepted the offer.

**Comment by johncschuster** [comment]
Ok everyone! I'm handling payments on [this issue](https://github.com/Expensify/App/issues/29702). I'm going to close this one out.
